xds/src/test/java/io/grpc/xds/CdsLoadBalancer2Test.java

```diff
@@ -458,6 +459,91 @@ public void aggregateCluster_intermediateClusterChanges() {
     assertThat(childBalancers).isEmpty();
   }
 
+  @Test
+  public void aggregateCluster_withLoops() {
+    String cluster1 = "cluster-01.googleapis.com";
+    // CLUSTER (aggr.) -> [cluster1]
+    CdsUpdate update =
+        CdsUpdate.forAggregate(CLUSTER, Collections.singletonList(cluster1))
+            .roundRobinLbPolicy().build();
+    xdsClient.deliverCdsUpdate(CLUSTER, update);
+    assertThat(xdsClient.watchers.keySet()).containsExactly(CLUSTER, cluster1);
+
+    // CLUSTER (aggr.) -> [cluster2 (aggr.)]
+    String cluster2 = "cluster-02.googleapis.com";
+    update =
+        CdsUpdate.forAggregate(cluster1, Collections.singletonList(cluster2))
+            .roundRobinLbPolicy().build();
+    xdsClient.deliverCdsUpdate(cluster1, update);
+    assertThat(xdsClient.watchers.keySet()).containsExactly(CLUSTER, cluster1, cluster2);
+
+    // cluster2 (aggr.) -> [cluster3 (EDS)]
+    String cluster3 = "cluster-03.googleapis.com";
+    CdsUpdate update2 =
+        CdsUpdate.forAggregate(cluster2, Arrays.asList(cluster3, cluster1, cluster2, cluster3))
+            .roundRobinLbPolicy().build();
+    xdsClient.deliverCdsUpdate(cluster2, update2);
+    assertThat(xdsClient.watchers.keySet()).containsExactly(CLUSTER, cluster1, cluster2, cluster3);
+    CdsUpdate update3 = CdsUpdate.forEds(cluster3, EDS_SERVICE_NAME, LRS_SERVER_INFO, 100L,
+        upstreamTlsContext, outlierDetection).roundRobinLbPolicy().build();
+    xdsClient.deliverCdsUpdate(cluster3, update3);
+    verify(helper).updateBalancingState(
+        eq(ConnectivityState.TRANSIENT_FAILURE), pickerCaptor.capture());
+    Status unavailable = Status.UNAVAILABLE.withDescription(
+        "CDS error: circular aggregate clusters directly under cluster-02.googleapis.com for root"
+            + " cluster cluster-foo.googleapis.com, named [cluster-01.googleapis.com,"
+            + " cluster-02.googleapis.com]");
+    assertPicker(pickerCaptor.getValue(), unavailable, null);
+  }
+
+  @Test
+  public void aggregateCluster_duplicateChildren() {
+    String cluster1 = "cluster-01.googleapis.com";
+    // CLUSTER (aggr.) -> [cluster1]
+    CdsUpdate update =
+        CdsUpdate.forAggregate(CLUSTER, Collections.singletonList(cluster1))
+            .roundRobinLbPolicy().build();
+    xdsClient.deliverCdsUpdate(CLUSTER, update);
+    assertThat(xdsClient.watchers.keySet()).containsExactly(CLUSTER, cluster1);
+
+    // CLUSTER (aggr.) -> [cluster2 (aggr.)]
+    String cluster2 = "cluster-02.googleapis.com";
+    update =
+        CdsUpdate.forAggregate(CLUSTER, Collections.singletonList(cluster2))
+            .roundRobinLbPolicy().build();
+    xdsClient.deliverCdsUpdate(CLUSTER, update);
+    assertThat(xdsClient.watchers.keySet()).containsExactly(CLUSTER, cluster2);
+
+    // cluster2 (aggr.) -> [cluster3 (EDS)]
+    String cluster3 = "cluster-03.googleapis.com";
+    CdsUpdate update2 =
```

Comment:
Why not reuse the "update" variable like before?

Reply:
I think that it is actually cleaner for each update to be in a different variable so you can see in the debugger exactly what has been done and it is easier to spot when you are accidentally forgetting to update the variable before delivering the update.